### PR TITLE
[Lang] Support kernel to return a matrix type value

### PR DESCRIPTION
**Pull request by mzmzm**
Related issue = #909

### CODE REVIEW

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 1]
```diff
@@ -1830,10 +1834,24 @@ class TaskCodegen : public IRVisitor {
...
     std::vector<std::tuple<spirv::SType, std::string, size_t>>
         struct_components_;
+    // Now we only have one ret
     for (auto &ret : ctx_attribs_->rets()) {
-      struct_components_.emplace_back(ir_->get_primitive_type(ret.dt),
-                                      "ret" + std::to_string(ret.index),
-                                      ret.offset_in_mem);
+      if (auto tensor_type = ret.dt->cast<TensorType>()) {
+        for (int i = 0; i < tensor_type->get_num_elements(); i++) {
+          struct_components_.emplace_back(
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:1841]: I don't quite understand this bit, wonder why there are two members for each element of the tensor?

Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp:1841]: Just for unifying the memory form with other backends. LLVM backend uses i64 data type to store return values, but Spir-v only support 32bit data type, so I use two i32 slots to represent 64bit.

File: taichi/backends/cc/codegen_cc.cpp [hunk 2]
```diff
@@ -187,8 +187,13 @@ class CCTransformer : public IRVisitor {
...
   void visit(ReturnStmt *stmt) override {
-    emit("ti_ctx->args[0].val_{} = {};",
-         data_type_name(stmt->element_types()[0]), stmt->values[0]->raw_name());
+    int idx{0};
+    for (auto &value : stmt->values) {
+      emit("ti_ctx->args[{}].val_{} = {};", idx++,
+           data_type_name(value->element_type()), value->raw_name());
+    }
+    // CC backend use arg buffer to access result values, so It doesn't support
+    // paas-by-ref now.
```
Comment (strongoier) [inline taichi/backends/cc/codegen_cc.cpp:196]: What does this comment mean?

File: tests/python/test_matrix_return.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+import pytest
+
+import taichi as ti
+from tests import test_utils
+
+### `ti.test`
+
+
+@test_utils.test(arch=ti.cpu)
```
[the remark this answers is not in the document]
Comment (strongoier) [reply tests/python/test_matrix_return.py:11]: BTW more tests are needed here (different return types, 30 return values).

File: taichi/backends/opengl/opengl_api.cpp [hunk 4]
```diff
@@ -222,7 +222,13 @@ bool initialize_opengl(bool use_gles, bool error_tolerance) {
...
 void CompiledTaichiKernel::init_args(Kernel *kernel) {
   arg_count = kernel->args.size();
-  ret_count = kernel->rets.size();
+  ret_count = 0;
+  for (auto &ret : kernel->rets) {
+    if (auto tensor_type = ret.dt->cast<TensorType>())
+      ret_count += tensor_type->get_num_elements();
+    else
+      ret_count += 1;
+  }
```
Comment (strongoier) [inline taichi/backends/opengl/opengl_api.cpp:231]: Is `ret_count` used anywhere?

File: taichi/backends/metal/kernel_manager.cpp [hunk 5]
```diff
@@ -510,7 +510,7 @@ class HostMetalCtxBlitter {
       char *device_ptr = base + ret.offset_in_mem;
       if (ret.is_array) {
         void *host_ptr = host_ctx_->get_arg<void *>(i);
```
Comment (strongoier) [inline taichi/backends/metal/kernel_manager.cpp:512]:
```suggestion
        void *host_ptr = host_ctx_->get_arg<void *>(i);
```
This should be redundant now.

File: taichi/backends/metal/codegen_metal.cpp [hunk 6]
```diff
@@ -429,7 +429,13 @@ class KernelCodegenImpl : public IRVisitor {
 
   void visit(ReturnStmt *stmt) override {
     // TODO: use stmt->ret_id instead of 0 as index
-    emit("*{}.ret0() = {};", kContextVarName, stmt->values[0]->raw_name());
+    int idx{0};
+    for (auto &value : stmt->values) {
+      emit("{}.ret0()[{}] = {};", kContextVarName, idx, value->raw_name());
+      idx += 2;
+      // metal only support i32 array, but there are i64 slots in
```
Comment (strongoier) [inline taichi/backends/metal/codegen_metal.cpp:436]: This comment is a bit confusing - "metal only support i32 array" is not true. We can only say `the return array of our Metal backend is of type i32`.

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 7]
```diff
@@ -848,11 +848,17 @@ class KernelGen : public IRVisitor {
...
     // TODO: use stmt->ret_id instead of 0 as index
-    emit("_args_{}_[{} >> {} + 0] = {};",
-         opengl_data_type_short_name(stmt->element_types()[0]),
-         taichi_opengl_ret_base,
-         opengl_data_address_shifter(stmt->element_types()[0]),
-         stmt->values[0]->short_name());
+    int idx{0};
+    for (auto &value : stmt->values) {
+      emit("_args_{}_[({} >> {}) + {}] = {};",
+           opengl_data_type_short_name(value->element_type()),
```
Comment (strongoier) [inline taichi/backends/opengl/codegen_opengl.cpp:854]: Is the `element_type()` here guaranteed to be `i32`?

File: taichi/ir/type_utils.cpp [hunk 8]
```diff
@@ -56,6 +56,11 @@ int data_type_size(DataType t) {
   //  setting a loud failure on pointers);
   //  2. Support pointer types here.
   t.set_is_pointer(false);
+  // If DataType is a TensorType, returns its element type size
```
Comment (strongoier) [inline taichi/ir/type_utils.cpp:59]: Why do we make the size of `TensorType` its element type size?

File: taichi/codegen/spirv/kernel_utils.cpp [hunk 9]
```diff
@@ -65,16 +65,36 @@ KernelContextAttributes::KernelContextAttributes(const Kernel &kernel)
...
     RetAttributes ra;
     ra.dt = kr.dt;
-    const size_t dt_bytes = data_type_size(ra.dt);
-    ra.is_array = false;  // TODO(#909): this is a temporary limitation
-    ra.stride = dt_bytes;
+    size_t dt_bytes{0};
+    if (auto tensor_type = ra.dt->cast<TensorType>()) {
+      dt_bytes = data_type_size(ra.dt);
+      ra.is_array = true;
+      ra.stride = tensor_type->get_num_elements() * dt_bytes * 2;
```
Comment (strongoier) [inline taichi/codegen/spirv/kernel_utils.cpp:72]: Why always *2? `dt_bytes` may not be 4.

**Review by strongoier — CHANGES_REQUESTED** [review]
Great work!

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 1 above)
Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:1841]: Hello, please generate the proper types. 64 bit data types are supported in SPIR-V.

Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:1841]: In tensor types, you should get the primitive type of the tensor, and generate a proper array type, then put the array type under the ret struct

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 10]
```diff
@@ -1828,10 +1834,18 @@ class TaskCodegen : public IRVisitor {
...
     std::vector<std::tuple<spirv::SType, std::string, size_t>>
         struct_components_;
+    // Now we only have one ret
     for (auto &ret : ctx_attribs_->rets()) {
-      struct_components_.emplace_back(ir_->get_primitive_type(ret.dt),
-                                      "ret" + std::to_string(ret.index),
-                                      ret.offset_in_mem);
+      if (auto tensor_type = ret.dt->cast<TensorType>()) {
+        struct_components_.emplace_back(
+            ir_->get_array_type(ir_->i64_type(),
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:1841]: Why always i64?

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 11]
```diff
@@ -516,18 +516,24 @@ class TaskCodegen : public IRVisitor {
...
+    }
     for (int i = 0; i < stmt->values.size(); i++) {
-      const auto &ret_attribs = ctx_attribs_->rets()[i];
-      const auto dt = ret_attribs.dt;
       spirv::Value buffer_val = ir_->make_value(
           spv::OpAccessChain,
           ir_->get_storage_pointer_type(ir_->get_primitive_type(dt)),
           get_buffer_value(BufferType::Rets, PrimitiveType::i32),
-          ir_->int_immediate_number(ir_->i32_type(), i));
+          ir_->int_immediate_number(ir_->i32_type(), (i << 1)));
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:530]: This statement is getting a pointer to the struct member variable. Should not shift here, but rather use the correct pointer types

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 10 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp:1841]: To match taichi's result buffer, now it only has 30 slots of uint64 type. For convenience, I try to cast the return value to i64.

File: taichi/backends/metal/kernel_manager.cpp (hunk 5 above)
Comment (mzmzm) [reply taichi/backends/metal/kernel_manager.cpp]: done

File: python/taichi/lang/ast/ast_transformer.py [hunk 12]
```diff
@@ -510,10 +511,24 @@ def build_Return(ctx, node):
...
+                if id(ctx.func.return_type) in primitive_types.type_ids:
+                    ctx.ast_builder.create_kernel_exprgroup_return(
+                        expr.make_expr_group(
+                            ti_ops.cast(expr.Expr(node.value.ptr),
+                                        ctx.func.return_type).ptr))
+                elif isinstance(ctx.func.return_type, MatrixType):
+                    ctx.ast_builder.create_kernel_exprgroup_return(
+                        expr.make_expr_group([
+                            ti_ops.cast(
+                                exp, ctx.func.return_type.dtype if isinstance(
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply python/taichi/lang/ast/ast_transformer.py]: done

File: taichi/ir/type.h [hunk 13]
```diff
@@ -174,7 +174,6 @@ class TensorType : public Type {
   std::vector<int> get_shape() const {
     return shape_;
   }
-
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/ir/type.h]: done

File: taichi/python/export_lang.cpp [hunk 14]
```diff
@@ -483,6 +488,8 @@ void export_lang(py::module &m) {
   py::class_<Kernel>(m, "Kernel")
       .def("get_ret_int", &Kernel::get_ret_int)
       .def("get_ret_float", &Kernel::get_ret_float)
+      .def("get_ret_int_matrix", &Kernel::get_ret_int_matrix)
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/python/export_lang.cpp]: done

File: taichi/python/export_lang.cpp [hunk 15]
```diff
@@ -386,8 +386,13 @@ void export_lang(py::module &m) {
...
            })
-      .def("decl_ret", [&](Program *program, const DataType &dt) {
-        return program->current_callable->insert_ret(dt);
+      .def("decl_ret",
+           [&](Program *program, const DataType &dt) {
+             return program->current_callable->insert_ret(dt);
+           })
+      .def("decl_tensor_type", [&](Program *program, std::vector<int> shape,
+                                   const DataType &element) {
+        return TypeFactory::create_tensor_type(shape, element);
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/python/export_lang.cpp]: done

File: taichi/backends/cc/codegen_cc.cpp (hunk 2 above)
Comment (mzmzm) [reply taichi/backends/cc/codegen_cc.cpp]: remove

File: taichi/backends/opengl/opengl_api.cpp (hunk 4 above)
Comment (mzmzm) [reply taichi/backends/opengl/opengl_api.cpp:231]: Yes, it was used to compute return buffer size. Although it is now unuseful in the current version of taichi.

File: taichi/backends/metal/codegen_metal.cpp (hunk 6 above)
Comment (mzmzm) [reply taichi/backends/metal/codegen_metal.cpp]: done

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 10 above)
Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:1843]: Maybe should consider using / modifying the host device contex blitter

File: taichi/ir/type_utils.cpp (hunk 8 above)
Comment (mzmzm) [reply taichi/ir/type_utils.cpp]: fix

File: taichi/backends/opengl/codegen_opengl.cpp (hunk 7 above)
Comment (mzmzm) [reply taichi/backends/opengl/codegen_opengl.cpp:854]: fix

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 11 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp]: fix

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 10 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp]: fix

File: taichi/codegen/spirv/kernel_utils.cpp (hunk 9 above)
Comment (mzmzm) [reply taichi/codegen/spirv/kernel_utils.cpp]: fix

File: taichi/python/export_lang.cpp [hunk 16]
```diff
@@ -873,7 +880,7 @@ void export_lang(py::module &m) {
 #if defined(TI_WITH_CUDA)
       return CUDAContext::get_instance().get_compute_capability();
 #else
-    TI_NOT_IMPLEMENTED
+          TI_NOT_IMPLEMENTED
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/python/export_lang.cpp]: fix

File: tests/python/test_matrix_return.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mzmzm) [reply tests/python/test_matrix_return.py]: added

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 1 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp]: Fix

File: taichi/ir/type_utils.h [hunk 17]
```diff
@@ -9,6 +9,8 @@ std::string data_type_name(DataType t);
 
 std::string data_type_format(DataType dt);
 
+int element_type_size(DataType t);
```
Comment (strongoier) [inline taichi/ir/type_utils.h:12]: Is this function used?

File: taichi/program/kernel.cpp [hunk 18]
```diff
@@ -319,7 +324,12 @@ float64 Kernel::get_ret_float(int i) {
 }
 
 int64 Kernel::get_ret_int(int i) {
-  auto dt = rets[i].dt->get_compute_type();
+  Type *dt;
+  if (auto tensor_type = rets[0].dt->cast<TensorType>()) {
```
Comment (strongoier) [inline taichi/program/kernel.cpp:328]: I think it's problematic to change `i` to 0 here. What about only abstracting the if-else for fetching results out?

File: taichi/transforms/type_check.cpp [hunk 19]
```diff
@@ -409,8 +409,16 @@ class TypeCheck : public IRVisitor {
 
   void visit(ReturnStmt *stmt) override {
     // TODO: Support stmt->ret_id?
-    stmt->ret_type = stmt->values[0]->ret_type;
-    TI_ASSERT(stmt->width() == 1);
+    if (stmt->get_kernel() != nullptr) {
```
Comment (strongoier) [inline taichi/transforms/type_check.cpp:412]: Assert this?

File: taichi/codegen/codegen_llvm.cpp [hunk 20]
```diff
@@ -1071,35 +1071,37 @@ void CodeGenLLVM::visit(ArgLoadStmt *stmt) {
...
-          builder->CreateBitCast(llvm_val[stmt->values[0]], intermediate_type),
-          dest_ty);
-      create_call("LLVMRuntime_store_result", {get_runtime(), extended});
-    } else {
-      TI_NOT_IMPLEMENTED;
+          builder->CreateBitCast(llvm_val[value], intermediate_type), dest_ty);
+      create_call(
+          "LLVMRuntime_store_result",
+          {get_runtime(), extended,
+           llvm::ConstantInt::get(*llvm_context, llvm::APInt(32, idx++))});
```
Comment (strongoier) [inline taichi/codegen/codegen_llvm.cpp:1104]: Use `tlctx->get_constant`?

File: taichi/backends/cc/cc_program.cpp [hunk 21]
```diff
@@ -41,7 +41,7 @@ void CCProgramImpl::materialize_snode_tree(
   layout_ = gen.compile();
   size_t root_size = layout_->compile();
   size_t gtmp_size = taichi_global_tmp_buffer_size;
-  size_t args_size = taichi_max_num_args * sizeof(uint64);
+  size_t args_size = taichi_result_buffer_entries * sizeof(uint64);
```
Comment (strongoier) [inline taichi/backends/cc/cc_program.cpp:44]: Why this change?

**Review by strongoier — CHANGES_REQUESTED** [review]

File: taichi/backends/cc/cc_program.cpp (hunk 21 above)
Comment (mzmzm) [reply taichi/backends/cc/cc_program.cpp:44]: Because the arg buffer and the result buffer are the same, and `result_buffer_entries` is greater than `max_num_args`

File: taichi/transforms/type_check.cpp (hunk 19 above)
Comment (mzmzm) [reply taichi/transforms/type_check.cpp:413]: fix

File: taichi/program/kernel.cpp (hunk 18 above)
Comment (mzmzm) [reply taichi/program/kernel.cpp]: fix

File: taichi/ir/type_utils.h (hunk 17 above)
Comment (mzmzm) [reply taichi/ir/type_utils.h]: removed

File: taichi/python/export_lang.cpp [hunk 22]
```diff
@@ -899,10 +901,16 @@ void export_lang(py::module &m) {
   m.def("get_type_factory_instance", TypeFactory::get_instance,
         py::return_value_policy::reference);
 
+  m.def("decl_tensor_type",
+        [&](std::vector<int> shape, const DataType &element) {
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/python/export_lang.cpp]: renamed

File: taichi/codegen/codegen_llvm.cpp (hunk 20 above)
Comment (mzmzm) [reply taichi/codegen/codegen_llvm.cpp]: fix

File: tests/python/test_matrix_return.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import pytest
+
+import taichi as ti
+from tests import test_utils
+
+### `ti.test`
+
+
+@test_utils.test()
+def test_arch():
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply tests/python/test_matrix_return.py:6]: modified

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 24]
```diff
@@ -516,13 +516,18 @@ class TaskCodegen : public IRVisitor {
...
+      dt = tensor_type->get_element_type();
+    }
     for (int i = 0; i < stmt->values.size(); i++) {
-      const auto &ret_attribs = ctx_attribs_->rets()[i];
-      const auto dt = ret_attribs.dt;
       spirv::Value buffer_val = ir_->make_value(
           spv::OpAccessChain,
           ir_->get_storage_pointer_type(ir_->get_primitive_type(dt)),
-          get_buffer_value(BufferType::Rets, PrimitiveType::i32),
+          get_buffer_value(BufferType::Rets, dt),
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:529]: Sorry I don't quite get it here. Why would this be able to return a matrix?

Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp:529]: Now we only support return one value, so if the return type is the matrix, its elements will be involved in `stmt->values`. Likewise, `ctx_attribs_->rets()`'s size will always be one. `dt` is the primitive type of the matrix element. And in the spir-v backend, I use a StructType to include an ArrayType, the matrix will be squeezed to an array when compiling. So to access the element, I change the argument of `OpAccessChain`, `0,i` instead of `i`.

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 25]
```diff
@@ -848,11 +848,20 @@ class KernelGen : public IRVisitor {
...
+           taichi_opengl_ret_base,
+           opengl_data_address_shifter(value->element_type()), idx,
+           value->short_name());
+      if (value->element_type()->is_primitive(PrimitiveTypeID::f64))
+        idx += 1;
+      else
+        idx += 2;
+      // opengl only support i32, f32 and f64 array, but there are 64bit slots
+      // in taichi's result buffer,so we need two slots to make them match.
```
Comment (strongoier) [inline taichi/backends/opengl/codegen_opengl.cpp:863]: Shall we wrap this hacky part to a function just like `opengl_data_address_shifter` above?

File: taichi/codegen/spirv/kernel_utils.cpp [hunk 26]
```diff
@@ -91,13 +97,47 @@ KernelContextAttributes::KernelContextAttributes(const Kernel &kernel)
     return bytes - offset;
   };
 
+  auto arange_rets = [](auto *vec, size_t offset) -> size_t {
+    std::vector<int> scalar_indices;
+    std::vector<int> array_indices;
+    for (int i = 0; i < vec->size(); ++i) {
```
Comment (strongoier) [inline taichi/codegen/spirv/kernel_utils.cpp:103]: Do we need to put arrays after scalars? `arange_args` doesn't do so here.

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 27]
```diff
@@ -516,13 +516,18 @@ class TaskCodegen : public IRVisitor {
   }
 
   void visit(ReturnStmt *stmt) override {
+    // Now we only support one ret
+    const auto &ret_attribs = ctx_attribs_->rets()[0];
+    auto dt = ret_attribs.dt;
+    if (auto tensor_type = dt->cast<TensorType>()) {
+      dt = tensor_type->get_element_type();
+    }
```
Comment (strongoier) [inline taichi/codegen/spirv/spirv_codegen.cpp:524]: Shall we get type directly from the `stmt`?

File: taichi/codegen/spirv/kernel_utils.cpp (hunk 26 above)
Comment (strongoier) [reply taichi/codegen/spirv/kernel_utils.cpp:103]: If we don't need that we can simplify the code in Metal backend as well.

File: taichi/transforms/type_check.cpp [hunk 28]
```diff
@@ -409,8 +409,17 @@ class TypeCheck : public IRVisitor {
 
   void visit(ReturnStmt *stmt) override {
     // TODO: Support stmt->ret_id?
-    stmt->ret_type = stmt->values[0]->ret_type;
     TI_ASSERT(stmt->width() == 1);
```
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/transforms/type_check.cpp:412]: Some tests don't have kernel, so it will get wrong on CI

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 24 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp:529]: For the scalar return type, I also use an ArrayType to involve this return value. So I can do the same operations when dealing with matrix and scalar.

File: taichi/transforms/type_check.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (mzmzm) [reply taichi/transforms/type_check.cpp:412]: fix

File: taichi/codegen/spirv/kernel_utils.cpp (hunk 26 above)
Comment (mzmzm) [reply taichi/codegen/spirv/kernel_utils.cpp]: fix

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 27 above)
Comment (mzmzm) [reply taichi/codegen/spirv/spirv_codegen.cpp]: fix

**Review by strongoier — APPROVED** [review]
LGTM!
